### PR TITLE
initial pass on refactor

### PR DESCRIPTION
**Pull request by whitej6**

### CODE REVIEW

File: nautobot_device_onboarding/jobs.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+from nautobot.extras.models import Role, SecretsGroup
+from nautobot.extras.choices import SecretsGroupSecretTypeChoices
+
+from nautobot_device_onboarding.netdev_keeper import NetdevKeeper
+
+
+PLUGIN_SETTINGS = settings.PLUGINS_CONFIG["nautobot_device_onboarding"]
+
+
+class OnboardingTask(Job):
```
Comment (jeffkala) [inline nautobot_device_onboarding/jobs.py:15]:
```suggestion
class OnboardingTask(Job):

    class Meta:
        """Meta object boilerplate for onboarding."""

        name = "Perform Device Onboarding"
        description = "Login to a device and populate Nautobot device object."
        has_sensitive_variables = False
```

File: nautobot_device_onboarding/jobs.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    credentials = ObjectVar(model=SecretsGroup, query_params={"": ""}, required=False, description="")
+    platform = ObjectVar(model=Platform, required=False, description="")
+    role = ObjectVar(model=Role, query_params={"": ""}, required=False, description="")
+    device_type = ObjectVar(model=DeviceType, required=False, description="")
+
+    def run(self, *args, **data):
+        """Process a single Onboarding Task instance."""
+        self.logger.info("START: onboard device")
+        credentials = self._parse_credentials(data["credentials"])
+        platform = data["platform"]
```
Comment (jeffkala) [inline nautobot_device_onboarding/jobs.py:36]: platform is not a required input to the job, does the key still come across with just a empty string?  Want to confirm we won't hit a KeyError here.

File: nautobot_device_onboarding/jobs.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+from nautobot_device_onboarding.netdev_keeper import NetdevKeeper
+
+
+PLUGIN_SETTINGS = settings.PLUGINS_CONFIG["nautobot_device_onboarding"]
+
+
+class OnboardingTask(Job):
+    location = ObjectVar(model=Location, query_params={"": ""}, required=False, description="")
+    ip_address = IPAddressVar(description="", label="")
+    port = IntegerVar(description="")
```
Comment (jeffkala) [inline nautobot_device_onboarding/jobs.py:25]: should we set a default? 22

File: nautobot_device_onboarding/jobs.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    timeout = IntegerVar(description="")
```
Comment (jeffkala) [inline nautobot_device_onboarding/jobs.py:26]: Set a default? 30 seconds?

File: nautobot_device_onboarding/jobs.py (hunk 3 above)
Comment (whitej6) [reply nautobot_device_onboarding/jobs.py:25]: All of that is pending filling in. Just wanted to commit before I signed off

Comment (jeffkala) [reply nautobot_device_onboarding/jobs.py:25]: Thanks! makes sense.  Will halt review for now.

File: nautobot_device_onboarding/jobs.py (hunk 2 above)
Comment (whitej6) [reply nautobot_device_onboarding/jobs.py:36]: Yep, just initial pass. Not ready for review yet just wanted to commit and open the PR for visibility

**Comment by jeffkala** [comment]
Did some testing tonight:
1. Can't get secret_group (credential) usage to work.
```
 {'exc_type': 'FieldError', 'exc_module': 'django.core.exceptions', 'exc_message': ["Cannot resolve keyword 'secret_type' into field. Choices are: _custom_field_data, created, description, destination_for_associations, id, last_updated, name, parameters, provider, secrets_group_associations, secrets_groups, source_for_associations, tagged_items, tags"]} 
 ```
 Would think it needs to be modified to something like
 ```python
         if credentials:
            try:
                self.username = credentials.get_secret_value(access_type = SecretsGroupAccessTypeChoices.TYPE_GENERIC, secret_type = SecretsGroupSecretTypeChoices.TYPE_USERNAME)
                self.password = credentials.get_secret_value(access_type = SecretsGroupAccessTypeChoices.TYPE_GENERIC, secret_type = SecretsGroupSecretTypeChoices.TYPE_PASSWORD)
                try:
                    secret = credentials.get_secret_value(access_type = SecretsGroupAccessTypeChoices.TYPE_GENERIC, secret_type = SecretsGroupSecretTypeChoices.TYPE_SECRET)
                except SecretsGroupAssociation.DoesNotExist:
                    self.secret = self.password
            except SecretsGroupAssociation.DoesNotExist:
                # should raise error or also fall back to NAPALM creds?

        else:
            self.username = settings.NAPALM_USERNAME
            self.password = settings.NAPALM_PASSWORD
            self.secret = settings.NAPALM_ARGS.get("secret", None)
```
If this is required we'd probably want to document the 'default" secret group access type to use.

2. I see odd authentication issues with authentication
```
 {'exc_type': 'OnboardException', 'exc_module': 'nautobot_device_onboarding.exceptions', 'exc_message': ['fail-general - fail-login - ERROR: Authentication to device failed.\n\nCommon causes of this problem are:\n1. Invalid username (redacted) password\n(redacted) Incorrect SSH-key file\n3. Connecting to the wrong device\n\nDevice settings: autodetect 10.1.1.8:22\n\n\nAuthentication failed: transport shut down or saw EOF']} 
 ```
 Can ssh from the server directly though. Still trying to figure that one out.
 Thinking its an issue with using the `settings.NAPALM_*` which assumes its in the settings, vs using os.getenv('NAPALM_*. Not sure which one is correct but find it as an antipattern compared to other plugins using those env vars.

The job does function properly when using env_vars instead of settings though.

File: nautobot_device_onboarding/nautobot_keeper.py [hunk 5]
```diff
@@ -34,8 +34,7 @@ def ensure_default_cf(obj, model):
         obj.validated_save()
     except ValidationError as err:
         raise OnboardException(
-            reason="fail-general",
-            message=f"ERROR: {obj} validation error: {err.messages}",
+            f"fail-general - ERROR: {obj} validation error: {err.messages}",
         ) from err
```
Comment (glennmatthews) [inline nautobot_device_onboarding/nautobot_keeper.py:38]: What's the reason for this global change to `OnboardException`? Was the more structured class not usable with a Job or something?

File: nautobot_device_onboarding/tests/test_api.py
Comment (glennmatthews) [inline nautobot_device_onboarding/tests/test_api.py:1]: Probably just delete this file altogether?

File: docs/user/faq.md [hunk 6]
```diff
@@ -28,38 +28,38 @@ PLUGINS_CONFIG = {
 
 ## How can I update the default credentials used to connect to a device?
 
-By default, the plugin uses the credentials defined in the main `nautobot_config.py` for NAPALM (`NAPALM_USERNAME`/`NAPALM_PASSWORD`). You can update the default credentials in `nautobot_config.py ` or you can provide specific one for each onboarding task.
+By default, the plugin uses the credentials defined in the main `nautobot_config.py` for NAPALM (`NAPALM_USERNAME`/`NAPALM_PASSWORD`/`DEVICE_SECRET`). You can update the default credentials in `nautobot_config.py ` or you can provide specific one for each onboarding job via a SecretsGroup.
```
[the remark this answers is not in the document]
Comment (whitej6) [reply docs/user/faq.md:31]: Napalm args, missed replacing that on the docs pass. will fix

File: nautobot_device_onboarding/jobs.py [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+PLUGIN_SETTINGS = settings.PLUGINS_CONFIG["nautobot_device_onboarding"]
+
+
+class OnboardingTask(Job):
+    """Nautobot Job for onboarding a new device."""
+
+    location = ObjectVar(
+        model=Location,
+        query_params={"content_type": "dcim.device"},
+        description="Name of parent Location for the onboarded device.",
```
[the remark this answers is not in the document]
Comment (whitej6) [reply nautobot_device_onboarding/jobs.py]: Copy pasta from old description.

File: nautobot_device_onboarding/nautobot_keeper.py (hunk 5 above)
Comment (whitej6) [reply nautobot_device_onboarding/nautobot_keeper.py:38]: Pickle error due so something in the custom exception

File: nautobot_device_onboarding/tests/test_api.py
Comment (whitej6) [reply nautobot_device_onboarding/tests/test_api.py:1]: Good catch

File: nautobot_device_onboarding/jobs.py [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+
+PLUGIN_SETTINGS = settings.PLUGINS_CONFIG["nautobot_device_onboarding"]
+
+
+class OnboardingTask(Job):
+    """Nautobot Job for onboarding a new device."""
+
+    location = ObjectVar(
+        model=Location,
+        query_params={"content_type": "dcim.device"},
```
[the remark this answers is not in the document]
Comment (scetron) [reply nautobot_device_onboarding/jobs.py:22]: Would explicit error handling also be good to have or is the error returned in this example enough to get a user on their way?

[the remark this answers is not in the document]
Comment (scetron) [reply nautobot_device_onboarding/jobs.py:22]: Also, while we automatically create a role if it's not there, might be good to point this out for roles as well.

File: nautobot_device_onboarding/migrations/0009_remove_models.py [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+        ),
+        migrations.RemoveField(
+            model_name="onboardingtask",
+            name="role",
+        ),
+        migrations.DeleteModel(
+            name="OnboardingDevice",
+        ),
+        migrations.DeleteModel(
+            name="OnboardingTask",
```
[the remark this answers is not in the document]
Comment (scetron) [reply nautobot_device_onboarding/migrations/0009_remove_models.py:36]: Definitely would be a good caveat, but I don't see users relying on the onboardingtasks long term

**Review by scetron — APPROVED** [review]
Overall, looks great!

File: nautobot_device_onboarding/jobs.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (whitej6) [reply nautobot_device_onboarding/jobs.py:22]: Role & location added to FAQ docs.

File: nautobot_device_onboarding/migrations/0009_remove_models.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (whitej6) [reply nautobot_device_onboarding/migrations/0009_remove_models.py:36]: Added to release notes that not only are all models being removed but all data from said models will be removed.

**Pull request merged**